lmdeploy/pytorch/engine/model_agent.py

```diff
@@ -628,18 +665,24 @@ def _broadcast_config(cache_config):
     try:
         config = model_config.hf_config
         torch_dtype = model_config.dtype
+        device_map = None
         with init_empty_weights():
             model = AutoModelForCausalLM.from_config(
                 config,
                 torch_dtype=torch_dtype,
                 trust_remote_code=trust_remote_code)
+            if rank == 0:
+                device_map = _create_device_map(model, world_size)
             _add_adapters(model, adapters)
+            if rank == 0:
+                # adapter would remove weight of linear.
+                device_map = _create_device_map(model, world_size, device_map)
         model.eval()
         model.config.use_cache = True
 
         if rank == 0:
             with LoadNoInit():
-                device_map = 'auto'
+                device_map = __get_device_map(model, device_map)
```

Comment:
Assuming there are two gpus and two linears, is the ﻿`device_map` on ﻿rank0 `{'linear0':0, 'linear1':1}`?

If the GPU memory of ﻿`linear0` + `﻿linear1` is greater than the available GPU memory of rank0, but `﻿linear0` and ﻿`linear1` are individually less than the available GPU memory of rank0 and rank1 respectively, will `﻿device_map` be switched to the CPU?

Reply:
> Assuming there are two gpus and two linears, is the ﻿device_map on ﻿rank0 {'linear0':0, 'linear1':1}?

Yes.

> will ﻿device_map be switched to the CPU?

t0:
- rank0: `linear0`
- rank1: `linear1`

t1
- rank0: `linear0/2`
- rank1: `linear0/2` + `linear1`

t2
- rank0: `linear0/2` + `linear1/2`
- rank1: `linear0/2` + `linear1/2`

Max memory usage would be  `linear0/2` + `linear1` on rank1

No CPU fallback.